src/plugins/data_source/server/saved_objects/credential_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,326 @@
+      );
+    };
+
+    const bulkCreateWithCredentialMaterialsEncryption = async <T = unknown>(
+      objects: Array<SavedObjectsBulkCreateObject<T>>,
+      options?: SavedObjectsCreateOptions
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      objects = await Promise.all(
+        objects.map(async (object) => {
+          const { type, attributes } = object;
+
+          if (this.type !== type) {
+            return object;
+          }
+
+          this.validateAttributes(attributes);
+
+          const { credentialMaterials } = attributes;
+
+          if (this.noAuthType === credentialMaterials.credentialMaterialsType) {
+            const noAuthAttributes = {
+              ...attributes,
+              credentialMaterials: this.dropCredentialMaterialsContent(credentialMaterials),
+            };
+            return {
+              ...object,
+              attributes: noAuthAttributes,
+            };
+          }
+
+          if (this.usernamePasswordType === credentialMaterials.credentialMaterialsType) {
+            const encryptedAttributes = {
+              ...attributes,
+              credentialMaterials: await this.validateAndEncryptCredentialMaterials(
+                credentialMaterials
+              ),
+            };
+
+            return {
+              ...object,
+              attributes: encryptedAttributes,
+            };
+          }
+
+          throw SavedObjectsErrorHelpers.createBadRequestError(
+            `Invalid credential materials type: '${credentialMaterials.credentialMaterialsType}'`
+          );
+        })
+      );
+      return await wrapperOptions.client.bulkCreate(objects, options);
+    };
+
+    const updateWithCredentialMaterialsEncryption = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (this.type !== type) {
+        return await wrapperOptions.client.update(type, id, attributes, options);
+      }
+
+      const { credentialMaterials } = attributes;
+
+      this.validateCredentialMaterials(credentialMaterials);
+
+      if (this.noAuthType === credentialMaterials.credentialMaterialsType) {
+        const noAuthAttributes: Partial<T> = {
+          ...attributes,
+          credentialMaterials: this.dropCredentialMaterialsContent(credentialMaterials),
+        };
+        return await wrapperOptions.client.update(type, id, noAuthAttributes, options);
+      }
+
+      if (this.usernamePasswordType === credentialMaterials.credentialMaterialsType) {
+        const encryptedAttributes: Partial<T> = {
+          ...attributes,
+          credentialMaterials: await this.validateAndEncryptCredentialMaterials(
+            credentialMaterials
+          ),
+        };
+        return await wrapperOptions.client.update(type, id, encryptedAttributes, options);
+      }
```

Comment:
Seems this similar code is being used in create/bulkCreate/update/bulkCreate can we move this to separate functions?

Reply:
Yes, it is not exact the same. For example, between create/update, we use different client operations as well as different attribute types (T VS PartialT>). The only comment part that has been extracted as functions are`validateAndEncryptCredentialMaterials` and `dropCredentialMaterialsContent `